File: Data Structures & Algorithms/course-schedule-ii/submission-11.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = {}

        for i in range(numCourses):
            adj[i] = []

        for course, pre in prerequisites:
            adj[course].append(pre)

        globalseen = set()
        res = []

        def dfs(course, seen):
            if course in seen:
                return False

            if course in globalseen:
                return True
                
            seen.add(course)

            for crs in adj[course]:
                if not dfs(crs, seen):
                    return False
            seen.remove(course)
            globalseen.add(course)
            res.append(course)

            return True
        
        for i in range(numCourses):
            seen = set()
            if i in res:
                continue
            if not i in globalseen: 
                if not dfs(i, seen):
                    return []

        return list(res)

File: Data Structures & Algorithms/course-schedule/submission-11.js
class Solution {
    /**
     * @param {number} numCourses
     * @param {number[][]} prerequisites
     * @return {boolean}
     */
    canFinish(numCourses, prerequisites) {
        let adj = new Map();

        for(let i = 0; i < numCourses; i++){
            adj.set(i, new Set());
        }

        for(let i = 0; i < prerequisites.length; i++){
            adj.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }

        //adj list
        let global = new Set();
        function dfs(seen, course){
            if(global.has(course)) return true;

            for(const c of adj.get(course)){
                if(seen.has(c)) return false;

                seen.add(c);
                if(!dfs(seen, c)) return false;
                seen.delete(c);
            }
            global.add(course);
            
            return true;

        }

        for(let i = 0; i < numCourses; i++){
            if(!dfs(new Set, i)) return false;
        }
        return true;
    }
}
